File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.cpp
#include <string>
#include <vector>

using namespace std;

bool checkSame(int x, int y, int size, const vector<vector<int>>& arr);
void dfs(int x, int y, int size, const vector<vector<int>>& arr);


int zeroCnt = 0;
int oneCnt = 0;

vector<int> solution(vector<vector<int>> arr) {
    dfs(0, 0, arr.size(), arr);
    return {zeroCnt, oneCnt};
}

bool checkSame(int x, int y, int size, const vector<vector<int>>& arr) {
    int val = arr[x][y];
    for(int i=x; i<x+size; i++) {
        for(int j=y; j<y+size; j++) {
            if(arr[i][j] != val) return false;
        }
    }
    
    return true;
}

void dfs(int x, int y, int size, const vector<vector<int>>& arr) {
    if(checkSame(x, y, size, arr)) {
        if(arr[x][y] == 0) zeroCnt++;
        else oneCnt++;
        return;
    }
    
    int half = size / 2;
    dfs(x, y, half, arr);
    dfs(x, y + half, half, arr);
    dfs(x + half, y, half, arr);
    dfs(x + half, y + half, half, arr);
}

File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.java
class Solution {
    static int zeroCnt = 0;
    static int oneCnt = 0;
    static int N;
    public int[] solution(int[][] arr) {
        N = arr.length;
        
        dfs(0, 0, N, arr);
        
        return new int[] {zeroCnt, oneCnt};
    }
    
    static void dfs(int r, int c, int size, int[][] arr) {
        if(checkSame(r, c, size, arr)) {
            if(arr[r][c] == 0) {
                zeroCnt++;
            } else {
                oneCnt++;
            }
            
            return;
        }
        
        int half = size / 2;
        dfs(r, c, half, arr);
        dfs(r, c + half, half, arr);
        dfs(r + half, c, half, arr);
        dfs(r + half, c + half, half, arr);
    }
    
    static boolean checkSame(int r, int c, int size, int[][] arr) {
        int value = arr[r][c];
        for (int i = r; i < r + size; i++) {
            for (int j = c; j < c + size; j++) {
                if (arr[i][j] != value) return false;
            }
        }
        return true;
    }
}